internal/sci/aws/server.go

```diff
@@ -0,0 +1,218 @@
+		Bucket: awsSdk.String(bucketName),
+		Key:    awsSdk.String(objectName),
+	}
+	_, err := s.Clients.S3Client.HeadObject(input)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() != s3.ErrCodeNoSuchKey {
+				return nil, fmt.Errorf("failed to head object: %w", err)
+			}
+		} else {
+			// It's an error of a different type, not an awserr.Error
+			return nil, fmt.Errorf("failed to head object: %w", err)
+		}
+	}
+	// Convert hex MD5 to base64
+	data, err := hex.DecodeString(checksum)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode MD5 checksum: %w", err)
+	}
+	base64md5 := base64.StdEncoding.EncodeToString(data)
+
+	reqInput := &s3.PutObjectInput{
+		Bucket:      awsSdk.String(bucketName),
+		Key:         awsSdk.String(objectName),
+		ContentType: awsSdk.String("application/octet-stream"),
+		ContentMD5:  awsSdk.String(base64md5),
+	}
+
+	expiration := time.Duration(req.GetExpirationSeconds()) * time.Second
+	putReq, _ := s.Clients.S3Client.PutObjectRequest(reqInput)
+	url, err := putReq.Presign(expiration)
+	if err != nil {
+		return nil, fmt.Errorf("failed to presign request: %w", err)
+	}
+	return &sci.CreateSignedURLResponse{Url: url}, nil
+}
+
+func (s *Server) BindIdentity(ctx context.Context, req *sci.BindIdentityRequest) (*sci.BindIdentityResponse, error) {
+	// Fetch the current trust policy
+	getRoleInput := &iam.GetRoleInput{
+		RoleName: awsSdk.String(req.Principal),
+	}
+	getRoleOutput, err := s.Clients.IamClient.GetRole(getRoleInput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the role: %v", err)
+	}
+
+	// URL decode the trust policy before decoding
+	decodedPolicy, err := url.QueryUnescape(*getRoleOutput.Role.AssumeRolePolicyDocument)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode trust policy: %v", err)
+	}
+
+	// Decode the current trust policy
+	var existingTrustPolicy map[string]interface{}
+	if err := json.Unmarshal([]byte(decodedPolicy), &existingTrustPolicy); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal trust policy: %v", err)
+	}
+
+	subValue := fmt.Sprintf("system:serviceaccount:%s:%s", req.KubernetesNamespace, req.KubernetesServiceAccount)
+
+	// Check if the OIDC provider's trust relationship already exists
+	statements := existingTrustPolicy["Statement"].([]interface{})
+	alreadyExists := false
+	for _, stmt := range statements {
+		stmtMap := stmt.(map[string]interface{})
+		if principal, ok := stmtMap["Principal"].(map[string]interface{}); ok {
+			if federated, ok := principal["Federated"].(string); ok && federated == s.OIDCProviderARN {
+				condition := stmtMap["Condition"].(map[string]interface{})["StringEquals"].(map[string]interface{})
+				condition[fmt.Sprintf("%s:sub", s.OIDCProviderURL)] = subValue
+				alreadyExists = true
+				break
+			}
+		}
+	}
+
+	// Construct the new trust relationship
+	newTrustRelationship := map[string]interface{}{
+		"Effect": "Allow",
+		"Principal": map[string]interface{}{
```

Comment:
could this be `map[string]string` ? I think so because `OIDCProviderARN` is of type `string`

Reply:
Principal is actually weirdly flexible as a string or an array of strings. I think the empty interface is the right way to handle that. This is valid:
```
"Principal": {
    "AWS": ["arn:aws:iam::123456789012:user/Bob", "arn:aws:iam::987654321098:root"]
}
Also valid:
```
"Principal": {
    "AWS": "arn:aws:iam::123456789012:root",
    "Service": "ec2.amazonaws.com",
    "Federated": "cognito-identity.amazonaws.com"
}
```
```